### DISCUSSION TITLE
Quanstions about entropy model and function update()

### DISCUSSION THREAD

**Discussion by AlbertoPresta** [InterDigitalInc/CompressAI#146] [Q&A]
Hi, 

sorry if I bother you with another question related to the entropy model calculated in the function update().

1. Why the quantized CDF is larger by two with respect to the cmf? is this because the first and the last values are not part of the CDF and are equal to 0 and 2^precision -1
2. When I update a pre-trained network (and also a network that I trained in my own dataset) the value of the tail_mass calculated in the function update() is equal to one, is it right? it seems quite weird to me.
3. the way the samples are calculated (samples = samples[None, :] + pmf_length[:, None, None]) is to move around the most common value of the latent specific dimension? 

 
thanks in advance for the answers,
Alberto

**Comment by YodaEmbedding** [comment]
1. Two reasons:
   - The cdf is indexed at "bin" boundaries, whereas each element of pmf represents the bins themselves.
    The probability of a symbol `x` is recovered via `pmf[x] = cdf[x + 1] - cdf[x]`.
    https://github.com/InterDigitalInc/CompressAI/blob/a29c3d564ce88a7ffd98b87b18c719766ecc47cd/compressai/entropy_models/entropy_models.py#L204-L212
    https://github.com/InterDigitalInc/CompressAI/blob/a29c3d564ce88a7ffd98b87b18c719766ecc47cd/compressai/cpp_exts/ops/ops.cpp#L71
   - One value needs to be allocated to represent all the low probability symbols outside the pmf range, where the "bypass" coding engine kicks in to try and encode those out-of-range symbols. The probability of this bypass symbol is roughly equal to the `tail_mass` (but not exactly equal, due to rounding precision).
     https://github.com/InterDigitalInc/CompressAI/blob/a29c3d564ce88a7ffd98b87b18c719766ecc47cd/compressai/cpp_exts/rans/rans_interface.cpp#L140-L147
2. This initializes it to `1e-9`:
   https://github.com/InterDigitalInc/CompressAI/blob/a29c3d564ce88a7ffd98b87b18c719766ecc47cd/compressai/entropy_models/entropy_models.py#L346
  For the `EntropyBottleneck`, it is the sum of the masses of the outer left/right regions of the logistic distribution:
   https://github.com/InterDigitalInc/CompressAI/blob/a29c3d564ce88a7ffd98b87b18c719766ecc47cd/compressai/entropy_models/entropy_models.py#L424
  For the `GaussianConditional`, it is the sum of the masses of the outer left/right regions of the normal distribution:
   https://github.com/InterDigitalInc/CompressAI/blob/a29c3d564ce88a7ffd98b87b18c719766ecc47cd/compressai/entropy_models/entropy_models.py#L642
3. Samples calculation.
  For the `EntropyBottleneck`, it looks like the additional dimension is needed due to the K-fold "matrix multiplication" for the small entropy modelling network described in the appendix of Balle2018:
   https://github.com/InterDigitalInc/CompressAI/blob/a29c3d564ce88a7ffd98b87b18c719766ecc47cd/compressai/entropy_models/entropy_models.py#L410-L423
   https://github.com/InterDigitalInc/CompressAI/blob/a29c3d564ce88a7ffd98b87b18c719766ecc47cd/compressai/entropy_models/entropy_models.py#L436-L455
  For the `GaussianConditional`, no extra dimension is needed since there's no extra matrix-based network needed and one can model the normal distribution directly:
   https://github.com/InterDigitalInc/CompressAI/blob/a29c3d564ce88a7ffd98b87b18c719766ecc47cd/compressai/entropy_models/entropy_models.py#L632-L640

**Comment by AlbertoPresta** [comment]
first of all thanks for the precious and precise answers. 
I got all the points you explained...but in practice, something weird happens ( or at least something that I can't understand fully):
I downloaded a pretrained model with the following command:

net = bmshj2018_factorized(quality=8, pretrained=True).eval().to(torch.device("cpu"))
net.update(force=True)

The I try to print out the quantized cdf and I have something like this (I show only the first dimension):

quantized_cdf = Tensor([ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9,
10, 11, 12, 13, 14, 15, 65536, 0, ...,0])
tail_mass = [1.]

Moreover, all dimensions have tail_mass equal to 1...what does it mean? should it equal to (or similar to) 1e-9? what is the domain of the quantized cdf if I want to plot it?

thanks in advance for the answers.
Alberto

**Reply by YodaEmbedding** [reply]
It might be that the model checkpoint wasn't loaded correctly. Could you please post a [MCVE](https://stackoverflow.com/help/minimal-reproducible-example) to help us verify this behavior?

**Reply by AlbertoPresta** [reply]
Yes, of course, I'll try to be as clear as possible: 

```python
from compressai.zoo import bmshj2018_factorized
import torch 

def upd(net):
    """
    this function simply reproduces what happens in the update function. 
    I wrote this function just to compute again the tail mass, the pmf and the quantized cdf
    """"

    medians = net.entropy_bottleneck.quantiles[:, 0, 1]

    minima = medians - net.entropy_bottleneck.quantiles[:, 0, 0]
    minima = torch.ceil(minima).int()
    minima = torch.clamp(minima, min=0)

    maxima = net.entropy_bottleneck.quantiles[:, 0, 2] - medians
    maxima = torch.ceil(maxima).int()
    maxima = torch.clamp(maxima, min=0)

    pmf_start = medians - minima
    pmf_length = maxima + minima + 1
        

    max_length = pmf_length.max().item()
    device = pmf_length.device
    samples = torch.arange(max_length, device=device)

    samples = samples[None, :] + pmf_length[:, None, None]
        
    half = float(0.5)
        
    lower = net.entropy_bottleneck._logits_cumulative(samples - half, stop_gradient=True)
    upper = net.entropy_bottleneck._logits_cumulative(samples + half, stop_gradient=True)
    sign = -torch.sign(lower + upper)
    pmf = torch.abs(torch.sigmoid(sign * upper) - torch.sigmoid(sign * lower))

    pmf = pmf[:, 0, :]
    tail_mass = torch.sigmoid(lower[:, 0, :1]) + torch.sigmoid(-upper[:, 0, -1:])
    quantized_cdf = net.entropy_bottleneck._pmf_to_cdf(pmf, tail_mass, pmf_length, max_length)
    
    return quantized_cdf, tail_mass




ml = torch.device("cpu")
net = bmshj2018_factorized(quality=1, pretrained=True).eval().to(ml)
net.update(force = True)

quantized_cdf, tail_mass = upd(net) # the outputs are like the ones I wrote in the previous answer

```

**Reply by YodaEmbedding** [reply]
Note that there are two lines that are different between `upd()` and the original `EntropyBottleneck.update()`:

```diff
-    device = pmf_length.device
+    device = pmf_start.device
     samples = torch.arange(max_length, device=device)
-    samples = samples[None, :] + pmf_length[:, None, None]
+    samples = samples[None, :] + pmf_start[:, None, None]
```

With these fixes, `0 <= tail_mass <= 6.2472e-10`.
